### PR TITLE
Refactor definition and response generator to properly handle slice, map and interface types

### PR DESCRIPTION
**Pull request by anilsenay**
I open PR to my old PR to show only additions to support all types including interfaces.

I have been burn out while doing this because there are many cases and it got more complicated 😄 Fixing something brokes another thing etc. And I am tired 😮‍💨

I think I achieved to support many types that you can see in [example/fiber/server.go](https://github.com/domhoward14/swagno/pull/5/files#diff-c74bf7bbf69e9b1bcad39e835df0536772b01217b2652250e043d10f9cdd8e1d)

But it came with complexity ofc. I think it should be refactored later.

I added hashes after each model which generated from its value. This provides overriding types expecially maps and interfaces. 

Btw, I will review my code and refactor necessary parts later in my free time. Can you point the parts which are complex when you review?

### CODE REVIEW

File: components/endpoint/endpoints.go [hunk 1]
```diff
@@ -110,8 +112,9 @@ func (e *EndPoint) GetPath() string {
 // GetBodyJsonParameter makes the body definitions and parameter for body if present. Parameters for body are described via schema
 // definition so that's why it doesn't use the 'Parameter' object like the other ones.
 func (e *EndPoint) GetBodyJsonParameter() *parameter.JsonParameter {
+	hash := utils.GenerateHash(utils.ConvertToString(e.Body))
 	if e.Body != nil {
-		bodyRef := fmt.Sprintf("#/definitions/%T", e.Body)
+		bodyRef := fmt.Sprintf("#/definitions/%T_%s", e.Body, hash)
```
Comment (domhoward14) [inline components/endpoint/endpoints.go:117]: Hmmm... I'm not a huge fan of the fact of when the client creates an object that when the definition of it renders on the UI it appends a hash that the client has no control over. Plus this solution looks like it'll add a hash even if the user creates an object without doing it the `response.New()` way, which is unecessary since the custom type created by the client is already unique.

I think if we just added a `TypeName` argument when the client calls `WithBody(body interface{})` then we could use that to have distinct names without creating and appending hashes.

File: components/endpoint/endpoints.go [hunk 2]
```diff
@@ -120,7 +123,7 @@ func (e *EndPoint) GetBodyJsonParameter() *parameter.JsonParameter {
 			bodySchema = parameter.JsonResponseSchema{
 				Type: "array",
 				Items: &parameter.JsonResponseSchemeItems{
-					Ref: fmt.Sprintf("#/definitions/%T", e.Body),
+					Ref: strings.ReplaceAll(fmt.Sprintf("#/definitions/%T_%s", e.Body, hash), "[]", ""),
```
Comment (domhoward14) [inline components/endpoint/endpoints.go:126]: Yeah, I don't think users want to see random hashes on the model names when the swagger UI renders the definitions.

File: example/fiber/server.go [hunk 3]
```diff
@@ -22,44 +21,35 @@ func main() {
...
+				response.New([]models.StructResponse1{}, "207", "Product List"),                                      // OK
+				response.New([]*models.StructResponse1{}, "208", "Product List"),                                     // OK
+				response.New([]map[string]string{}, "209", "Product List"),                                           // OK
+				response.New([]interface{}{}, "210", "Product List"),                                                 // OK
+				response.New([]interface{}{models.StructResponse1{}}, "211", "Product List"),                         // not desired result
+				response.New([][]models.StructResponse1{}, "212", "Product List"),                                    // OK
+				response.New([][]*models.StructResponse1{}, "213", "Product List"),                                   // OK
+				response.New(map[string]int{"field": 123}, "214", "Product List"),                                    // OK
+				response.New(map[string]*int{}, "215", "Product List"),                                               // OK
+				response.New(map[string]models.StructResponse1{}, "216", "Product List"),                             // OK
```
Comment (domhoward14) [inline example/fiber/server.go:41]: Do we need to support hashmaps and structs? I'm thinking that one would be sufficient, since they both encode to json key-value pairs.

File: generator/definition.go [hunk 4]
```diff
@@ -18,29 +19,59 @@ func NewDefinitionGenerator(definitionMap map[string]definition.Definition) *Def
 		Definitions: definitionMap,
 	}
 }
+func (g DefinitionGenerator) getDefiniationName(t interface{}) string {
```
Comment (domhoward14) [inline generator/definition.go:22]: Typo in the function name here

File: generator/definition.go [hunk 5]
```diff
@@ -18,29 +19,59 @@ func NewDefinitionGenerator(definitionMap map[string]definition.Definition) *Def
...
+	hash := utils.GenerateHash(utils.ConvertToString(t))
+	// create random names for maps
+	if reflect.TypeOf(t).Kind() == reflect.Map && reflect.TypeOf(t).Name() == "" {
+		return strings.ReplaceAll(fmt.Sprintf("%T_%s_%s", t, utils.GetHashOfMap(utils.ConvertInterfaceToMap(t)), hash), "[]", "")
+	}
+	return strings.ReplaceAll(fmt.Sprintf("%T_%s", t, hash), "[]", "")
+}
 
-func (g DefinitionGenerator) CreateDefinition(t interface{}) {
+func (g DefinitionGenerator) CreateDefinition(t interface{}, prefix string) {
```
Comment (domhoward14) [inline generator/definition.go:31]: Yeah, refactor would be nice here once the time comes. I recommend we use the 'Chain Of Responsibility' pattern for this. It follows a lot of the SOLID principles and would make this look a lot more cleaner and easy to test.

File: example/fiber/server.go [hunk 6]
```diff
@@ -22,44 +21,35 @@ func main() {
 			endpoint.WithMethod(endpoint.GET),
 			endpoint.WithPath("/product"),
 			endpoint.WithTags("product"),
-			endpoint.WithSuccessfulReturns([]response.Info{response.New([]models.Product{}, "200", "Product List")}),
+			endpoint.WithSuccessfulReturns([]response.Info{
+				response.New("string", "200", "Product List"),                                                        // OK
```
Comment (domhoward14) [inline example/fiber/server.go:25]: I see a lot of these have "OK" on their lines. Did you do manual testing?

File: generator/definition.go [hunk 7]
```diff
@@ -49,12 +80,19 @@ func (g DefinitionGenerator) CreateDefinition(t interface{}) {
...
-func (g DefinitionGenerator) createStructDefinitions(_struct reflect.Type) map[string]definition.DefinitionProperties {
+func (g DefinitionGenerator) createStructDefinitions(_struct reflect.Type, data any, prefix string) map[string]definition.DefinitionProperties {
 	properties := make(map[string]definition.DefinitionProperties)
 	for i := 0; i < _struct.NumField(); i++ {
 		field := _struct.Field(i)
 		fieldType := getType(field.Type.Kind().String())
 		fieldJsonTag := getJsonTag(field)
+		exampleTag := getExampleTag(field)
+
+		// check swagno tag for custom type
```
Comment (domhoward14) [inline generator/definition.go:91]: we have another json struct tag that we're parsing?

**Comment by domhoward14** [comment]
@anilsenay  I was thinking of merging in your changes but holding off on some like the `response.New()`, `CustomResponseType`, and the new supported types that you're parsing until we add some testing so we don't rush in to it or delay our release going down more rabbit holes of complexity. The initial set of functionality we were working on seems to be a good starting point, and the automated tests i have for it seems to be working great. What are your thoughts on that? Do you see any changes that I pointed out that you think are critical to be released in v1.0.0? Let me know when convenient 😁 !

File: components/endpoint/endpoints.go (hunk 2 above)
Comment (anilsenay) [reply components/endpoint/endpoints.go:126]: This is necessary to prevent overriding. If you have a structs which have interface field like: `Struct { Field interface{} }` then if you want to do something like `SomeStruct1{ Field: SomeStruct2}` and `SomeStruct1{ Field: SomeStruct3 }` the last one will override the first one because they both will have same definition name something like `model.SomeStruct`

Generating hash based on its values provides different struct names and there is no overriding. But this is the only solution came in my mind. I am not a fan of this solution for the reasons you already mentioned.

File: example/fiber/server.go (hunk 3 above)
Comment (anilsenay) [reply example/fiber/server.go:41]: I am not sure actually, I just wanted to make it more flexible. For the case maybe user return map in their handler function without struct etc.

File: example/fiber/server.go (hunk 6 above)
Comment (anilsenay) [reply example/fiber/server.go:25]: Yep I was testing in UI and checking them one by one. I will delete them, I want to show they works

File: generator/definition.go (hunk 7 above)
Comment (anilsenay) [reply generator/definition.go:91]: There was a pull request:
https://github.com/go-swagno/swagno/pull/18 

What do you think?

**Comment by anilsenay** [comment]
> @anilsenay I was thinking of merging in your changes but holding off on some like the `response.New()`, `CustomResponseType`, and the new supported types that you're parsing until we add some testing so we don't rush in to it or delay our release going down more rabbit holes of complexity. The initial set of functionality we were working on seems to be a good starting point, and the automated tests i have for it seems to be working great. What are your thoughts on that? Do you see any changes that I pointed out that you think are critical to be released in v1.0.0? Let me know when convenient 😁 !

Hmm, I think the last changes that I made is not production ready. I already did them for experimantally to see if we can achieve support more types. You see there is no test etc. Maybe we can just release v1.0.0 without this PR by just merging only your PR. In your PR the whole API changed but the generation logic did not effect so generation will be same as before without breaking anything. 

In summary,
- https://github.com/domhoward14/swagno/pull/4 will be merged
- https://github.com/domhoward14/swagno/pull/3 will be merged (i think #4 already covers this pr)
- (make last refactors after merge them)
- After this two pr merged, your pr https://github.com/go-swagno/swagno/pull/21 will be ready to merge
- After release v1.0.0, we can work on supporting new types in time. I tried to make it on time so It made the code a lot complex by adding bad written codes. I think we can do it better if we don't try to push 

What do you think? @domhoward14

File: components/endpoint/endpoints.go (hunk 1 above)
Comment (anilsenay) [reply components/endpoint/endpoints.go:117]: TypeName idea seems ok but I dont want expect user to know more API 😄 But we should definitely evaluate it too.

**Comment by domhoward14** [comment]
> > @anilsenay I was thinking of merging in your changes but holding off on some like the `response.New()`, `CustomResponseType`, and the new supported types that you're parsing until we add some testing so we don't rush in to it or delay our release going down more rabbit holes of complexity. The initial set of functionality we were working on seems to be a good starting point, and the automated tests i have for it seems to be working great. What are your thoughts on that? Do you see any changes that I pointed out that you think are critical to be released in v1.0.0? Let me know when convenient 😁 !
> 
> Hmm, I think the last changes that I made is not production ready. I already did them for experimantally to see if we can achieve support more types. You see there is no test etc. Maybe we can just release v1.0.0 without this PR by just merging only your PR. In your PR the whole API changed but the generation logic did not effect so generation will be same as before without breaking anything.
> 
> In summary,
> 
> * [Feat/refactoring #4](https://github.com/domhoward14/swagno/pull/4) will be merged
> * [Refactor definition and response #3](https://github.com/domhoward14/swagno/pull/3) will be merged (i think [Feat/refactoring #4](https://github.com/domhoward14/swagno/pull/4) already covers this pr)
> * (make last refactors after merge them)
> * After this two pr merged, your pr [v1.0.0 update go-swagno/swagno#21](https://github.com/go-swagno/swagno/pull/21) will be ready to merge
> * After release v1.0.0, we can work on supporting new types in time. I tried to make it on time so It made the code a lot complex by adding bad written codes. I think we can do it better if we don't try to push
> 
> What do you think? @domhoward14

Yup, totally agree! I think we're on the same page. I have a list of things i'm going to finish in the next 2 days then i'll be ready to merge `Feat/refactoring` into `v1.0.0` 😁
